packages/syft/src/syft/core/tensor/smpc/share_tensor.py

```diff
@@ -413,23 +423,98 @@ def apply_function(
         Returns:
             ShareTensor: Result of the operation.
         """
+        ShareTensor.sanity_check(y)
 
-        op = ShareTensor.get_op(self.ring_size, op_str)
-        numpy_type = utils.RING_SIZE_TO_TYPE.get(self.ring_size, None)
+        op = ShareTensor.get_op(x.ring_size, op_str)
+        numpy_type = utils.RING_SIZE_TO_TYPE.get(x.ring_size, None)
         if numpy_type is None:
-            raise ValueError(f"Do not know numpy type for ring size {self.ring_size}")
+            raise ValueError(f"Do not know numpy type for ring size {x.ring_size}")
 
         if isinstance(y, ShareTensor):
-            utils.get_ring_size(self.ring_size, y.ring_size)
-            value = op(self.child, y.child)
+            utils.get_ring_size(x.ring_size, y.ring_size)
+            value = op(x.child, y.child)
         else:
-            # TODO: Converting y to numpy because doing "numpy op torch tensor" raises exception
-            value = op(self.child, np.array(y, numpy_type))  # TODO: change to np.int64
+            if op_str in {"add", "sub"}:
+                # TODO: Converting y to numpy because doing "numpy op torch tensor" raises exception
+                value = (
+                    op(x.child, np.array(y, numpy_type))
+                    if x.rank == 0
+                    else deepcopy(x.child)
+                )
+            elif op_str == "mul":
+                value = op(x.child, np.array(y, numpy_type))
+            else:
+                raise ValueError(f"{op_str} not supported")
 
-        res = self.copy_tensor()
+        res = x.copy_tensor()
         res.child = value
         return res
 
+    def apply_action_function(self, op_str: str) -> Optional[ShareTensor]:
+        """Apply action based operations.
+
+        Args:
+            y (Union[int, float, torch.Tensor, np.ndarray, "ShareTensor"]): tensor to apply the operator.
+            op_str (str): Operator.
+
+        Returns:
+            ShareTensor: Result of the operation.
+        """
+        # relative
+        from ...node.common.action import smpc_action_functions
+
+        # TODO: refactor in next iteration, creation of action, accesses the DB again
+        # we could pass in the sharetensor values directly from this function.
+
+        nr_parties = self.nr_parties
+        rank = self.rank
+        method_name = op_str
+
+        actions_generator = smpc_action_functions.get_action_generator_from_op(
+            operation_str=method_name, nr_parties=nr_parties
+        )
+
+        args_id = context.GLOBAL_ARGS_ID
```

Comment:
Yeah...I am not a fun of using this context in this way -- because we could pass this information at the level of the function - `apply_action_function` - in my mind the `context` should be kept only for reducing the information which remains the same/get updated when doing SMPC -- like the random number generators

Reply:
Temporarily removed actions.